util-collection/src/main/scala/sbt/internal/util/Attributes.scala

```diff
@@ -11,9 +11,27 @@ import Types._
 import scala.reflect.ClassTag
 import sbt.util.OptJsonWriter
 
-// T must be invariant to work properly.
+enum KeyTag[A]:
+  case Setting[A](typeArg: Class[?]) extends KeyTag[A]
+  case Task[A](typeArg: Class[?]) extends KeyTag[A]
+  case InputTask[A](typeArg: Class[?]) extends KeyTag[A]
```

Comment:
Eventually `typeArg` would be a `String`. But we need to fix the `maybeAddOutputsAndFileStamps` method first.